### PR TITLE
Upon a slot being filled in a form, added ability to run a given action.

### PR DESCRIPTION
**Pull request by ivanmkc**
**Proposed changes**:
I wanted to be able to perform an action after a slot is filled in a form. 

Demo:

```
2021-04-18 16:40:40 INFO     root  - Rasa server is up and running.
Bot loaded. Type a message and press enter (use '/stop' to exit): 
Your input ->  Hi I want my id                                                                                                                                                                              
What's your name?
Your input ->  Ivan                                                                                                                                                                                                 
Nice to meet you Ivan!   <- This is the new action
What's your hometown?
Your input ->                
```

The NLU for forms will have the additional optional property for each slot called `slot_filled_action`:
```
forms:
  student_id_card_form:
    name:
      - type: from_entity
        entity: PERSON
        slot_filled_action: "action_confirm_name_slot_filled"
```

The first `slot_filled_action` action will be used.

This isn't ideal but since each slot is represented by a list and not a dictionary, I couldn't easily add a property without refactoring the format representing the slot.

Better proposal: Allow list of potential responses that will be matched depending on what slots are available (like how DialogFlow works). The first matching response_action will be fired. This would require changing some of the validation logic around forms.
```
forms:
  student_id_card_form:
    name:
      - type: from_entity
        entity: PERSON
    birthday:
      - type: from_entity
        entity: DATE
    response_actions: 
      - action_confirm_name_slot_filled
      - action_confirm_birthday_slot_filled
      - action_confirm_name_and_birthday_slots_filled
```

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @rctatman will take a look at it as soon as possible ✨

File: rasa/core/actions/forms.py [hunk 1]
```diff
@@ -520,7 +520,44 @@ async def validate(
...
+
+            if len(requested_slot_mappings) > 0:
+                slot_filled_action_name = None
+                for mapping in requested_slot_mappings:
+                    slot_filled_action_name = mapping.get("slot_filled_action")
+
+                    if slot_filled_action_name:
+                        break
+
+                if slot_filled_action_name:
```
Comment (ivanmkc) [inline rasa/core/actions/forms.py:539]: TODO: Fix issue where if an "other" slot is filled, this will call the wrong action.

Comment (ivanmkc) [reply rasa/core/actions/forms.py:558]: Fixed.

File: rasa/core/actions/forms.py [hunk 2]
```diff
@@ -334,7 +334,10 @@ def get_slot_to_fill(self, tracker: "DialogueStateTracker") -> Optional[str]:
         )
 
     def extract_requested_slot(
-        self, tracker: "DialogueStateTracker", domain: Domain, slot_to_fill: Text,
```
Comment (ivanmkc) [inline rasa/core/actions/forms.py:337]: I think black auto-formatted this.

File: rasa/core/actions/forms.py [hunk 3]
```diff
@@ -749,22 +799,25 @@ async def is_done(
         # We explicitly check only the last occurrences for each possible termination
         # event instead of doing `return event in events_so_far` to make it possible
         # to override termination events which were returned earlier.
-        return next(
```
Comment (ivanmkc) [inline rasa/core/actions/forms.py:752]: More black auto-formatting

File: rasa/core/actions/forms.py [hunk 4]
```diff
@@ -520,7 +523,54 @@ async def validate(
...
+            tracker, validation_events, domain
+        )
+
+        slots_filled = [
+            event.key
+            for event in validation_events
+            if isinstance(event, SlotSet) and not event.key == REQUESTED_SLOT
+        ]
+
+        for slot_filled in slots_filled:
```
Comment (ivanmkc) [inline rasa/core/actions/forms.py:542]: The order is governed by the validation events. Perhaps better to use the order of the slots defined in the form.

**Pull request closed without merging**